compiler/test/dotty/tools/dotc/core/tasty/CommentPicklingTest.scala

```diff
@@ -0,0 +1,151 @@
+import java.nio.file.attribute.BasicFileAttributes
+import java.util.EnumSet
+
+import scala.collection.JavaConverters.asScalaIteratorConverter
+import scala.concurrent.duration.{Duration, DurationInt}
+
+import org.junit.Test
+import org.junit.Assert.{assertEquals, assertFalse, fail}
+
+class CommentPicklingTest {
+
+  val compileOptions = TestConfiguration.defaultOptions and "-Ykeep-comments" and "-Yemit-tasty"
+  val unpickleOptions = TestConfiguration.defaultOptions
+
+  @Test def commentOnDef: Unit = {
+    val sources = "object A { /** foo */ def bar = 2 }" :: Nil
+    compileAndCheckComment(sources, "bar".toTermName, Some("/** foo */"))
+  }
+
+  @Test def commentOnVal: Unit = {
+    val sources = "object A { /** foo */ val bar = 2 }" :: Nil
+    compileAndCheckComment(sources, "bar".toTermName, Some("/** foo */"))
+  }
+
+  @Test def commentLocalVal: Unit = {
+    val sources = "object A { def buzz = { /** foo */ val bar = 3 } }" :: Nil
+    compileAndCheckComment(sources, "bar".toTermName, Some("/** foo */"))
+  }
+
+  @Test def commentLocalDef: Unit = {
+    val sources = "object A { def buzz = { /** foo */ def bar = 5 } }" :: Nil
+    compileAndCheckComment(sources, "bar".toTermName, Some("/** foo */"))
+  }
+
+  @Test def commentOnClass: Unit = {
+    val sources = "/** foo */ class A" :: Nil
+    compileAndCheckComment(sources, "A".toTypeName, Some("/** foo */"))
+  }
+
+  @Test def commentOnObject: Unit = {
+    val sources = "/** foo */ object A" :: Nil
+    compileAndCheckComment(sources, "A".toTermName, Some("/** foo */"))
+  }
+
+  @Test def commentOnlazyVal: Unit = {
+    val sources = "class A { /** foo */ lazy val buzz = 2 }" :: Nil
+    compileAndCheckComment(sources, "buzz".toTermName, Some("/** foo */"))
+  }
+
+  private def compileAndCheckComment(sources: List[String], treeName: Name, expectedComment: Option[String]): Unit = {
+    compileAndUnpickle(sources) { (trees, ctx) =>
+      findTreeNamed(treeName)(trees, ctx) match {
+        case Some(md: tpd.MemberDef) =>
+          val symbol = md.symbol(ctx)
+          val comment = for { docCtx <- ctx.docCtx
+                              comment <- docCtx.docstring(symbol) } yield comment.raw
+          assertEquals(expectedComment, comment)
+        case other =>
+          fail(s"Unexpected: $other")
+      }
+    }
+  }
+
+  private def findTreeNamed(name: Name)(trees: List[tpd.Tree], ctx: Context): Option[tpd.NameTree] = {
+    val acc = new tpd.TreeAccumulator[Option[tpd.NameTree]] {
+      override def apply(x: Option[tpd.NameTree], tree: tpd.Tree)(implicit ctx: Context): Option[tpd.NameTree] = {
+        x.orElse(tree match {
+          case md: tpd.NameTree if md.name == name => Some(md)
+          case other => foldOver(None, other)
+        })
+      }
+    }
+    acc(None, trees)(ctx)
+  }
+
+  private def compileAndUnpickle[T](sources: List[String])(fn: (List[tpd.Tree], Context) => T) = {
+    inTempDirectory { tmp =>
```

Comment:
A nice to have (that shouldn't block this PR): a driver that compiles from virtual files. I think it would be fairly easy: override `doCompile` and call `compileSources(List[SourceFile])` instead of  `compile(List[String])`. Creating a virtual source file is straightforward: `new SourceFile("<fake name>", content.toCharArray)`